### PR TITLE
Support any process writing to cache invalidation stream.

### PR DESCRIPTION
**Pull request by erikjohnston**
This is done via two big changes:
1. Change the cache stream to use the new `MultiWriterIdGenerator`
2. Change the definition of `Stream.current_token` to accept an instance name parameter, which allows us to support the idea of there being multiple writers and so multiple positions to keep track of.

Its probably worth going through commit by commit, as there is a bit of boiler plate refactoring going on.

### CODE REVIEW

**Comment by clokep** [comment]
@erikjohnston Looks like this has a conflict!

**Comment by erikjohnston** [comment]
> @erikjohnston Looks like this has a conflict!

But it didn't just now :sob:

File: synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+ /* for some reason, we have accumulated duplicate entries in
+  * device_lists_outbound_pokes, which makes prune_outbound_device_list_pokes less
+  * efficient.
+  */
```
Comment (clokep) [inline synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres:19]: Is this a copy-paste from a different file? Doesn't look related to this change.

Comment (erikjohnston) [reply synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres:19]: Yeah, I think I must have mentally parsed it as part of the license blurb

File: synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+DROP TABLE IF EXISTS cache_invalidation_stream;
+
+CREATE TABLE cache_invalidation_stream (
```
Comment (clokep) [inline synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres:23]: I don't know if we care, but re-using this table name means it will be hard to downgrade.

Comment (erikjohnston) [reply synapse/storage/data_stores/main/schema/delta/58/cache_instance.sql.postgres]: Good point, I've renamed it and truncated the old table.
